pkg/stanza/fileconsumer/file.go

```diff
@@ -211,24 +214,22 @@ func (m *Manager) makeFingerprint(path string) (*fingerprint.Fingerprint, *os.Fi
 // makeReader take a file path, then creates reader,
 // discarding any that have a duplicate fingerprint to other files that have already
 // been read this polling interval
-func (m *Manager) makeReaders(paths []string) []*reader.Reader {
-	readers := make([]*reader.Reader, 0, len(paths))
-OUTER:
+func (m *Manager) makeReaders(paths []string) {
+	m.activeFiles.Clear()
```

Comment:
Generally, clearing out this file set without a complementary action of moving the files somewhere may be problematic. If it's actually clearing out any files, it should be placing them in previous. If it's not, then we don't need it. I think it's better if we handle this in one place, at the end of the poll.

---

I would actually like to suggest we move to a different paradigm with these file sets. In this implementation, we're creating the file sets and then treating them as containers which stay in place, and then all file management is done by removing files from one set and adding them to another.

I propose that we would be better off _rotating the sets themselves_. This requires some explanation though.

First, I acknowledge we are taking steps in the direction of managing open files more flexibly. However, I think we need to set those ambitions aside for just a bit longer and simplify the current implementation first. To that end, we _currently_ have three sets:
1. active (open)
2. previous (open)
3. known (closed)

One of the immediate benefits of moving to file sets is we can simplify the way we handle `knownFiles`. This is intended to represent "about 3 or 4 previous poll intervals worth of files", and this is implemented as a single set with a moving size. Instead, we can now move to `[]fileset` with a fixed length and remove all the moving size tracking/calculating code. Doing this will mean we effectively have the following sets:

1. active (open)
2. previous (1 poll ago - open)
3. known[0] (2 polls ago - closed)
4. known[1] (3 polls ago - closed)
5. known[2] (4 polls ago - closed) 

Then, in many cases we can just rotate the sets at the end of every poll: 
- drop known[2]
- known[1] -> known[2]
- known[0] -> known[1]
- close previous files and create a set based on metadata
- active -> previous
- new active

This should simplify the code, including the file set interface. We shouldn't need `Clear`, `Reset` or `PopN`. I also think it will be more performant.

There other notable exception to this model is when we're searching back through our history to `Match` a file. In that case, we should still remove the match and add it to the active set. I think this is a clear exception to the pattern though. WDYT?

Reply:
When you say "3 or 4 polls before", do you mean we should get rid of moving averages? And shift to our previous implementation(in which we kept record for past 2 polls only)?

Reply:
> get rid of moving averages

Yes, this was only meant as a temporary fix. See https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/28567#issuecomment-1777408974

I'm ok with _not_ removing it in this PR, but think with the intention in mind, it highlights the usefulness of rotating file sets. If we don't remove it now, I'll likely open a PR to do so very soon.